core/string/translation.cpp

```diff
@@ -1159,9 +1159,9 @@ StringName TranslationServer::_get_message_from_translations(const StringName &p
 
 		StringName r;
 		if (!plural) {
-			r = t->get_message(p_message, p_context);
+			r = t->call("get_message", p_message, p_context);
 		} else {
-			r = t->get_plural_message(p_message, p_message_plural, p_n, p_context);
+			r = t->call("get_plural_message", p_message, p_message_plural, p_n, p_context);
```

Comment:
In general we do not allow scripts to override methods from the engine classes. The common pattern for these cases is to register a virtual method for scripting that can be overridden, and on the engine side the implementation (in this case `get_message()` and `get_plural_message()`) should call the virtual method if there's a script attached.

Reply:
As an example, BaseButton has a `_pressed()` virtual method that will be called when the button is pressed:

https://github.com/godotengine/godot/blob/382ddd497a732812356fccbe9642d86848a4da9d/scene/gui/base_button.cpp#L123-L129

Note that the fact that the C++ function having the same name as the virtual function is just coincidence, since the C++ function isn't bound to scripting only the virtual version is:

https://github.com/godotengine/godot/blob/382ddd497a732812356fccbe9642d86848a4da9d/scene/gui/base_button.cpp#L426

Reply:
You can find more examples of this kind of virtual scripting methods in #48746 as many have been renamed to harmonize their names.

So here you should likely bind `_get_message()` and `_get_plural_message()` as virtual methods additionally to the C++ functions `get_message()` and `get_plural_message()` (which are also public and bound IINM). Then the implementation of `get_message()` would call `_get_message()` if defined in the script:

```gdscript
StringName Translation::get_message(const StringName &p_src_text, const StringName &p_context) const {
	if (get_script_instance()) {
		return get_script_instance()->call("_get_message", p_src_text, p_context);
	}

	if (p_context != StringName()) {
		WARN_PRINT("Translation class doesn't handle context. Using context in get_message() on a Translation instance is probably a mistake. \nUse a derived Translation class that handles context, such>
	}

	const Map<StringName, StringName>::Element *E = translation_map.find(p_src_text);
	if (!E) {
		return StringName();
	}

	return E->get();
}
```

Reply:
In this case you should also check if the `_get_message()` function is defined in the script.

Reply:
Rebased and changed code to call script instance methods.

Reply:
Looks good, but it still needs actually binding these as virtual methods as documented here: https://github.com/godotengine/godot/pull/44670#discussion_r662508489
Otherwise users won't know they exist, and then they serve little purpose. This will also add them to the documentation so you can fill their descriptions.

Reply:
As I understand BIND_VMETHOD supposed to be removed in https://github.com/godotengine/godot/pull/51983

Reply:
Ah you're right, but just removing it is not sufficient, it needs to be replaced by the new syntax. Search e.g. `_get_connection_line` in #51983 for an example of what's needed.

I suggest splitting this in `master` and `3.x` PRs as we'll probably want the ability to override `get_message` in `3.x`, but there it should use `BIND_VMETHOD` and `call` like you do here.

Reply:
Should I open PR with `BIND_VMETHOD` on `3.x` branch?

Reply:
Yes, please do :)

Reply:
Changed this PR to `GDVIRTUAL_BIND`.